### PR TITLE
Add playlist name and video name in playlist sharing content

### LINKED ISSUE #9025: Playlist sharing: include video names & playlist name
### Checklist

- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have read the [FAQ](https://newpipe.net/FAQ/) and my problem isn't listed.
- [X] I'm aware that this is a request for NewPipe itself and that requests for adding a new service need to be made at [NewPipeExtractor](https://github.com/TeamNewPipe/NewPipeExtractor/issues).
- [X] I have taken the time to fill in all the required details. I understand that the feature request will be dismissed otherwise.
- [X] This issue contains only one feature request.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Feature description

In shared text the video name and playlist name need to be shown.

### Why do you want this feature?

Its very annoying click in all links to find a specific video.

### Additional information

The current feature (that shares just the urls) can be a "Copy to clipboard" button to easy manipulation

But to share to people, just the links is extremely hard to find a video. Like: in 100 videos find a specific video about something.

### PR DESCRIPTION
**Pull request by snaik20**
#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

Add playlist name and video name in playlist sharing content

- Currently, only a list of videos separated by newline are added in
  the share content.
- This makes it difficult to identify a specific video in a list of
  Urls.
- Used string resources for the sharing content formats.
- Added a confirmation dialog for users to choose between sharing
  playlist formats.
- Added Playlist name as the header and corresponding video name for
  each video Url in following format.
```
My Playlist
- Music1: https://media-url1
- Music2: https://media-url2
- Music3: https://media-url3
```

#### Before/After Screenshots/Screen Record

| Before | After | Confirmation Dialog |
| --- | --- | --- |
|  |  |  |

#### Video demo:
https://github.com/TeamNewPipe/NewPipe/assets/87667048/09962034-146a-4fa6-acbc-5bc58cba2d10

#### Fixes the following issue(s)

- Supersedes and closes #9060
- Fixes #9025 

#### APK testing

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).
**Note:** I am aware of the NewPipe rewrite, and I would be open to updating my PR as per requirements.

### CODE REVIEW

**Comment by TobiGr** [comment]
Thank you for the changes. Please add screenshots of the share sheets showing the state before and after your PR.

**Comment by snaik20** [comment]
@TobiGr Thanks for your inputs, updated the PR with screenshots.

**Comment by TobiGr** [comment]
Thanks!
Sometimes, people just need to URLs to be exported. Please add a preference to the ContentSettings which allows the user to export a list of URLs only.

**Comment by TobiGr** [comment]
@Stypox just commented on Matrix that a dialog before sharing is better than a setting. He is correct. So please do not add a setting but rather an AlertDialog with a checkmark which enables the user add the video title and playlist name

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -811,4 +811,10 @@
     <string name="channel_tab_about">About</string>
     <string name="show_channel_tabs">Channel tabs</string>
     <string name="show_channel_tabs_summary">What tabs are shown on the channel pages</string>
+    <string name="share_playlist">Share Playlist</string>
+    <string name="share_playlist_with_details_message">Share playlist with details such as playlist name and video titles Or a simply share a list of video links</string>
+    <string name="share_playlist_with_details">Share with details</string>
+    <string name="share_playlist_with_list">Share list</string>
```
Comment (TobiGr) [inline app/src/main/res/values/strings.xml:817]:
```suggestion
    <string name="share_playlist_with_details_message">Share playlist with details such as playlist name and video titles or as a simple list of video URLs</string>
    <string name="share_playlist_with_details">Share with details</string>
    <string name="share_playlist_with_list">Share URL list</string>
```

File: app/src/main/res/values/settings_keys.xml [hunk 2]
```diff
@@ -24,6 +24,8 @@
     <string name="screen_brightness_timestamp_key">screen_brightness_timestamp_key</string>
     <string name="clear_queue_confirmation_key">clear_queue_confirmation_key</string>
     <string name="ignore_hardware_media_buttons_key">ignore_hardware_media_buttons_key</string>
+    <string name="share_playlist_with_details_key">share_playlist_with_details_key</string>
+    <string name="share_playlist_with_details_can_show_dialog_key">share_playlist_with_details_can_show_dialog_key</string>
```
Comment (TobiGr) [inline app/src/main/res/values/settings_keys.xml:28]: I think this name is better to understand what the key is used for. I think the usage needs to be inverted though.
```suggestion
    <string name="remember_local_playlist_sharing_option">remember_local_playlist_sharing_option</string>
```

File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java [hunk 3]
```diff
@@ -374,16 +385,27 @@ public boolean onOptionsItemSelected(final MenuItem item) {
...
+    private void sharePlaylist() {
+        final boolean shouldSharePlaylistDetails = sharedPreferences.getBoolean(
+                requireContext().getString(R.string.share_playlist_with_details_key), false);
         disposables.add(playlistManager.getPlaylistStreams(playlistId)
                 .flatMapSingle(playlist -> Single.just(playlist.stream()
                         .map(PlaylistStreamEntry::getStreamEntity)
-                        .map(StreamEntity::getUrl)
+                        .map(streamEntity -> {
+                            if (shouldSharePlaylistDetails) {
+                                return String.format("- %s: %s",
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:398]: I think we will need to allow the localization of this string for LTR countries. Please extract it into a string res.

File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java [hunk 4]
```diff
@@ -841,5 +863,40 @@ private PlayQueue getPlayQueue(final int index) {
...
+                .setCancelable(true)
+                .setView(dialogBinding.getRoot())
+                .setPositiveButton(R.string.share_playlist_with_details, (dialog, which) -> {
+                    sharedPreferences.edit()
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_can_show_dialog_key),
+                                    !dialogBinding.rememberChoiceCheckBox.isChecked())
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_key), true)
+                            .commit();
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:886]:
```suggestion
                            .apply();
```

File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java [hunk 5]
```diff
@@ -841,5 +863,40 @@ private PlayQueue getPlayQueue(final int index) {
...
+                    sharePlaylist();
+                })
+                .setNegativeButton(R.string.share_playlist_with_list, (dialog, which) -> {
+                    sharedPreferences.edit()
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_can_show_dialog_key),
+                                    !dialogBinding.rememberChoiceCheckBox.isChecked())
+                            .putBoolean(requireContext().getString(
+                                    R.string.share_playlist_with_details_key), false)
+                            .commit();
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java:896]:
```suggestion
                            .apply();
```

File: app/src/main/res/xml/video_audio_settings.xml [hunk 6]
```diff
@@ -249,5 +249,13 @@
...
             app:singleLineTitle="false"
             app:iconSpaceReserved="false" />
+
+        <SwitchPreferenceCompat
+            android:defaultValue="false"
+            android:key="@string/share_playlist_with_details_key"
+            android:summary="@string/share_playlist_with_details_settings_item"
+            android:title="@string/share_playlist"
+            app:singleLineTitle="false"
+            app:iconSpaceReserved="false" />
```
Comment (TobiGr) [inline app/src/main/res/xml/video_audio_settings.xml:259]: I am not sure if this is the correct place for this setting. @opusforlife2 Where would you put it?

**Comment by snaik20** [comment]
@TobiGr Added a confirmation dialog to the playlist sharing process, allowing users to choose between sharing playlist formats. Users can also choose to save their sharing format preference to skip the confirmation dialog. In addition, an option has been added to the settings to update the sharing preference.

**Review by TobiGr — COMMENTED** [review]
Thank you

File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java (hunk 4 above)
Comment (snaik20) [reply app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java]: I had used `commit` as we need to make the changes synchronously for the `sharePlaylist` method. I have  tested with `apply` and it seems to work properly.

**Comment by snaik20** [comment]
Thanks @TobiGr for your time and review. Updated the changes as requested. Let us wait for @opusforlife2 on inputs for settings item.

**Comment by snaik20** [comment]
> There's no good place to put it, unfortunately. `Content` menu is our current catch-all for miscellaneous settings. (If someone would like to tackle #9587 it would make everyone's lives easier. _wink wink_)
> 
> But why do we need a setting for this at all? Just have the checkbox state remembered. That one extra dialogue doesn't make much of a difference.

Hi @opusforlife2,

Thanks for your feedback.

I understand your concern about the extra dialog, but I believe it's important to give users the option to choose how they want to share their playlists. Some users may prefer the convenience of having a checkbox that remembers their preference, while others may want to be more deliberate about how they share their playlists.

That said, I'm open to other suggestions. Here are a few options:
 - We could remove the setting item and display the dialog every time a user shares a playlist.
 - We could remove the dialog and only keep the setting, and let the user know where to find the setting item to change the default behavior.

I'm happy to discuss these options further with you, or to implement any other solution that you think would be better.

Regarding #9587, I agree that it would be great if someone could tackle it. I would be happy to look into it.

Thanks

**Comment by snaik20** [comment]
https://github.com/TeamNewPipe/NewPipe/assets/87667048/5ac5ab6f-163d-4407-bd67-a078a8b55bbb



> > every time a user shares a playlist
> 
> I can't think of any obvious use cases where an average user would need to rapidly go through playlists and share them. At best this would be a one-off task. Apart from that, there could be an occasional desire to share an interesting playlist with a friend. This is why I suggested to go with:
> 
> > We could remove the setting item and display the dialog every time
> 
> Can you think of any important use cases that are frequent enough to justify an additional setting?

Hi @opusforlife2,

I think you make a good point. I can't think of any use cases where the average user would need to rapidly share playlists. I have updated the flow to always show the dialog and updated the video demo.

WDYT

**Comment by snaik20** [comment]
@opusforlife2 Thanks, updated.

**Review by TobiGr — APPROVED** [review]
Thanks!

**Comment by TobiGr** [comment]
I fixed the wrongly formatted string resources in 0758cd6980d2cda88803930d34227b11f10d05d2

**Comment by snaik20** [comment]
> I fixed the wrongly formatted string resources in 0758cd6980d2cda88803930d34227b11f10d05d2

Thanks @TobiGr

**Pull request merged**